File: apps/web/src/lib/campaign-calendar.ts
"use client";

import { useCallback, useEffect, useSyncExternalStore } from "react";

export type CampaignCalendarApiErrorCode =
  "unauthorized" | "forbidden" | "not_found" | "validation" | "network_failure";

export class CampaignCalendarApiError extends Error {
  constructor(
    readonly code: CampaignCalendarApiErrorCode,
    message: string,
    readonly status?: number,
    options?: ErrorOptions,
  ) {
    super(message, options);
    this.name = "CampaignCalendarApiError";
  }
}

export type CampaignCalendarEventType =
  | "campaign.start"
  | "campaign.target_end"
  | "campaign.milestone.target"
  | "marketing.content.scheduled"
  | "marketing.content.channel_scheduled"
  | "marketing.content.published"
  | "marketing.content.channel_published"
  | "marketing.content.approval_requested"
  | "marketing.content.approved";

export type CampaignCalendarSourceType =
  | "campaign"
  | "campaign_milestone"
  | "marketing_content_item"
  | "marketing_content_channel"
  | "approval_request"
  | string;

export type CampaignCalendarCampaignContext = {
  id: string;
  name: string;
  status: string;
  campaign_type: string;
};

export type CampaignCalendarArtistContext = {
  id: string;
  name: string;
};

export type CampaignCalendarReleaseContext = {
  id: string;
  title: string;
  artist_id: string | null;
};

export type CampaignCalendarChannelContext = {
  id: string;
  channel: string;
  placement: string;
};

export type CampaignCalendarApprovalContext = {
  request_id: string | null;
  state: string | null;
  label: string | null;
  approved_revision_is_current: boolean | null;
  can_schedule: boolean | null;
  available_actions: string[];
};

export type CampaignCalendarEvent = {
  id: string;
  event_type: CampaignCalendarEventType | string;
  source_type: CampaignCalendarSourceType;
  source_id: string;
  source_parent_id: string | null;
  title: string;
  description: string | null;
  starts_at: string;
  ends_at: string | null;
  date: string | null;
  all_day: boolean;
  timezone: string;
  status: string | null;
  campaign: CampaignCalendarCampaignContext | null;
  artist: CampaignCalendarArtistContext | null;
  release: CampaignCalendarReleaseContext | null;
  channel: CampaignCalendarChannelContext | null;
  approval: CampaignCalendarApprovalContext | null;
  url: string | null;
  sort_key: string;
};

export type CampaignCalendarResponse = {
  workspace_id: string;
  start: string;
  end: string;
  timezone: string;
  events: CampaignCalendarEvent[];
  total: number;
  limit: number;
  offset: number;
};

export type CampaignCalendarQueryOptions = {
  start?: string | null;
  end?: string | null;
  timezone?: string | null;
  campaign_id?: string | null;
  artist_id?: string | null;
  release_id?: string | null;
  status?: string | null;
  event_types?: string | readonly string[] | null;
  include_archived?: boolean | null;
  include_published?: boolean | null;
  limit?: number | null;
  offset?: number | null;
};

export type CampaignCalendarResourceState<T> = {
  data: T | null;
  error: CampaignCalendarApiError | null;
  isLoading: boolean;
  isMutating: boolean;
  reload: () => Promise<T>;
};

type CacheEntry<T> = {
  data: T | null;
  error: CampaignCalendarApiError | null;
  fetcher: (() => Promise<T>) | null;
  isLoading: boolean;
  listeners: Set<() => void>;
  promise: Promise<T> | null;
  version: number;
};

const cache = new Map<string, CacheEntry<unknown>>();

export const campaignCalendarQueryKeys = {
  all: "campaign-calendar",
  workspaceRange: (workspaceId: string, options?: CampaignCalendarQueryOptions) =>
    `campaign-calendar:workspace-range:${workspaceId}:${stableCampaignCalendarQueryKey(options)}`,
};

function definedQueryEntries(options?: CampaignCalendarQueryOptions) {
  return Object.entries(options ?? {}).filter(
    ([, value]) => value !== undefined && value !== null && value !== "",
  );
}

function normalizedEventTypes(value: string | readonly string[]): string[] {
  const values = typeof value === "string" ? value.split(",") : value;
  return values
    .map((item) => item.trim())
    .filter(Boolean)
    .sort();
}

export function stableCampaignCalendarQueryKey(
  options?: CampaignCalendarQueryOptions | null,
): string {
  if (!options) {
    return "default";
  }
  const key = definedQueryEntries(options)
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([key, value]) => {
      if (key === "event_types" && (Array.isArray(value) || typeof value === "string")) {
        return `${key}:${normalizedEventTypes(value).join(",")}`;
      }
      return `${key}:${String(value)}`;
    })
    .join("|");
  return key || "default";
}

export function campaignCalendarQuery(options?: CampaignCalendarQueryOptions): string {
  const params = new URLSearchParams();
  for (const [key, value] of definedQueryEntries(options)) {
    if (key === "event_types" && (Array.isArray(value) || typeof value === "string")) {
      for (const item of normalizedEventTypes(value)) {
        params.append(key, item);
      }
      continue;
    }
    params.set(key, String(value));
  }
  const query = params.toString();
  return query ? `?${query}` : "";
}

function entryFor<T>(key: string): CacheEntry<T> {
  let entry = cache.get(key) as CacheEntry<T> | undefined;
  if (!entry) {
    entry = {
      data: null,
      error: null,
      fetcher: null,
      isLoading: false,
      listeners: new Set(),
      promise: null,
      version: 0,
    };
    cache.set(key, entry as CacheEntry<unknown>);
  }
  return entry;
}

function emit(entry: CacheEntry<unknown>) {
  entry.version += 1;
  for (const listener of entry.listeners) {
    listener();
  }
}

function errorDetailMessage(detail: unknown): string | null {
  if (typeof detail === "string") {
    return detail;
  }
  if (Array.isArray(detail)) {
    return detail
      .map((entry) => {
        if (entry && typeof entry === "object" && "msg" in entry) {
          return String(entry.msg);
        }
        return null;
      })
      .filter(Boolean)
      .join(" ");
  }
  return null;
}

async function responseErrorDetail(response: Response): Promise<string | null> {
  try {
    const payload = (await response.json()) as { detail?: unknown };
    return errorDetailMessage(payload.detail);
  } catch {
    return null;
  }
}

function toCampaignCalendarApiError(
  status: number,
  detail?: string | null,
): CampaignCalendarApiError {
  if (status === 401) {
    return new CampaignCalendarApiError(
      "unauthorized",
      "Sign in again to load the campaign calendar.",
      status,
    );
  }
  if (status === 403) {
    return new CampaignCalendarApiError(
      "forbidden",
      "You do not have access to the campaign calendar.",
      status,
    );
  }
  if (status === 404) {
    return new CampaignCalendarApiError("not_found", "Campaign calendar was not found.", status);
  }
  if (status === 400 || status === 422) {
    return new CampaignCalendarApiError(
      "validation",
      detail || "Campaign calendar filters have validation errors.",
      status,
    );
  }
  return new CampaignCalendarApiError(
    "network_failure",
    "Campaign calendar could not be loaded.",
    status,
  );
}

async function campaignCalendarJson<T>(path: string, init?: RequestInit): Promise<T> {
  const headers = new Headers(init?.headers);
  headers.set("Accept", "application/json");

  let response: Response;
  try {
    response = await fetch(path, { ...init, cache: "no-store", headers });
  } catch (error) {
    throw new CampaignCalendarApiError(
      "network_failure",
      "Unable to reach the campaign calendar API.",
      undefined,
      { cause: error },
    );
  }

  if (!response.ok) {
    throw toCampaignCalendarApiError(response.status, await responseErrorDetail(response));
  }
  return (await response.json()) as T;
}

function normalizeCampaignCalendarResponse(
  payload: CampaignCalendarResponse,
): CampaignCalendarResponse {
  return {
    ...payload,
    events: payload.events.map((event) => ({
      ...event,
      source_parent_id: event.source_parent_id ?? null,
      description: event.description ?? null,
      ends_at: event.ends_at ?? null,
      date: event.date ?? null,
      status: event.status ?? null,
      campaign: event.campaign ?? null,
      artist: event.artist ?? null,
      release: event.release ?? null,
      channel: event.channel ?? null,
      approval: event.approval
        ? {
            ...event.approval,
            available_actions: event.approval.available_actions ?? [],
          }
        : null,
      url: event.url ?? null,
    })),
  };
}

async function loadResource<T>(key: string, fetcher: () => Promise<T>): Promise<T> {
  const entry = entryFor<T>(key);
  entry.fetcher = fetcher;
  if (entry.promise) {
    return entry.promise;
  }

  entry.error = null;
  entry.isLoading = true;
  emit(entry);

  entry.promise = fetcher()
    .then((data) => {
      entry.data = data;
      entry.error = null;
      return data;
    })
    .catch((error) => {
      entry.error =
        error instanceof CampaignCalendarApiError
          ? error
          : new CampaignCalendarApiError(
              "network_failure",
              "Campaign calendar could not be loaded.",
              undefined,
              { cause: error },
            );
      throw entry.error;
    })
    .finally(() => {
      entry.isLoading = false;
      entry.promise = null;
      emit(entry);
    });

  emit(entry);
  return entry.promise;
}

function useCampaignCalendarResource<T>(
  key: string | null,
  fetcher: (() => Promise<T>) | null,
): CampaignCalendarResourceState<T> {
  const subscribe = useCallback(
    (listener: () => void) => {
      if (!key) {
        return () => undefined;
      }
      const entry = entryFor<T>(key);
      entry.listeners.add(listener);
      return () => {
        entry.listeners.delete(listener);
      };
    },
    [key],
  );

  const getSnapshot = useCallback(() => (key ? entryFor<T>(key).version : 0), [key]);
  useSyncExternalStore(subscribe, getSnapshot, getSnapshot);

  useEffect(() => {
    if (!key || !fetcher) {
      return;
    }
    const entry = entryFor<T>(key);
    entry.fetcher = fetcher;
    if (entry.data === null && !entry.isLoading) {
      void loadResource(key, fetcher).catch(() => undefined);
    }
  }, [fetcher, key]);

  const reload = useCallback(async () => {
    if (!key || !fetcher) {
      throw new CampaignCalendarApiError(
        "not_found",
        "A campaign calendar resource key is required.",
      );
    }
    return loadResource(key, fetcher);
  }, [fetcher, key]);

  const entry = key ? entryFor<T>(key) : null;
  return {
    data: entry?.data ?? null,
    error: entry?.error ?? null,
    isLoading: entry?.isLoading ?? false,
    isMutating: false,
    reload,
  };
}

export function invalidateCampaignCalendarCache(predicate?: (key: string) => boolean) {
  for (const [key, entry] of cache.entries()) {
    if (predicate && !predicate(key)) {
      continue;
    }
    entry.data = null;
    entry.error = null;
    if (entry.fetcher) {
      void loadResource(key, entry.fetcher).catch(() => undefined);
    } else {
      emit(entry);
    }
  }
}

export function invalidateCampaignCalendarWorkspaceCache(workspaceId: string) {
  invalidateCampaignCalendarCache((key) =>
    key.startsWith(`campaign-calendar:workspace-range:${workspaceId}:`),
  );
}

export function clearCampaignCalendarCache() {
  cache.clear();
}

export function listCampaignCalendarEvents(
  workspaceId: string,
  options: CampaignCalendarQueryOptions,
): Promise<CampaignCalendarResponse> {
  return campaignCalendarJson<CampaignCalendarResponse>(
    `/api/workspaces/${workspaceId}/campaign-calendar${campaignCalendarQuery(options)}`,
  ).then(normalizeCampaignCalendarResponse);
}

export function useCampaignCalendar(
  workspaceId: string | null,
  options?: CampaignCalendarQueryOptions,
): CampaignCalendarResourceState<CampaignCalendarResponse> {
  const key = workspaceId ? campaignCalendarQueryKeys.workspaceRange(workspaceId, options) : null;
  const fetcher = useCallback(
    () => listCampaignCalendarEvents(workspaceId ?? "", options ?? {}),
    [options, workspaceId],
  );
  return useCampaignCalendarResource(key, workspaceId ? fetcher : null);
}

File: apps/api/src/labelos_api/api/v1/campaign_calendar.py
from __future__ import annotations

from dataclasses import asdict
from datetime import datetime
from typing import Annotated, NoReturn
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from labelos_api.auth import CurrentUserContext, SessionDep, get_current_user_context
from labelos_api.services import campaign_calendar_service
from labelos_api.services.campaign_calendar_service import (
    MAX_CAMPAIGN_CALENDAR_LIMIT,
    CampaignCalendarAuthorizationError,
    CampaignCalendarEventQuery,
    CampaignCalendarValidationError,
    NormalizedCampaignCalendarEvent,
)

router = APIRouter(prefix="/workspaces", tags=["campaign-calendar"])


class CampaignCalendarCampaignContextResponse(BaseModel):
    id: UUID
    name: str
    status: str
    campaign_type: str


class CampaignCalendarArtistContextResponse(BaseModel):
    id: UUID
    name: str


class CampaignCalendarReleaseContextResponse(BaseModel):
    id: UUID
    title: str
    artist_id: UUID | None


class CampaignCalendarChannelContextResponse(BaseModel):
    id: UUID
    channel: str
    placement: str


class CampaignCalendarApprovalContextResponse(BaseModel):
    request_id: UUID | None
    state: str | None
    label: str | None
    approved_revision_is_current: bool | None
    can_schedule: bool | None
    available_actions: list[str]


class CampaignCalendarEventResponse(BaseModel):
    id: str
    event_type: str
    source_type: str
    source_id: UUID
    source_parent_id: UUID | None
    title: str
    description: str | None
    starts_at: datetime
    ends_at: datetime | None
    date: str | None
    all_day: bool
    timezone: str
    status: str | None
    campaign: CampaignCalendarCampaignContextResponse | None
    artist: CampaignCalendarArtistContextResponse | None
    release: CampaignCalendarReleaseContextResponse | None
    channel: CampaignCalendarChannelContextResponse | None
    approval: CampaignCalendarApprovalContextResponse | None
    url: str | None
    sort_key: str


class CampaignCalendarResponse(BaseModel):
    workspace_id: UUID
    start: datetime
    end: datetime
    timezone: str
    events: list[CampaignCalendarEventResponse]
    total: int
    limit: int
    offset: int


def _not_found() -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")


def _forbidden(detail: str) -> HTTPException:
    return HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=detail)


def _bad_request(detail: str) -> HTTPException:
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=detail)


def _raise_capability_denial(reason: str) -> NoReturn:
    if reason in {"invalid_resource_scope", "membership_not_found"}:
        raise _not_found()
    if reason == "insufficient_department_access":
        raise _forbidden("Insufficient department access")
    raise _forbidden("Insufficient capability permission")


def _service_error(
    exc: CampaignCalendarAuthorizationError | CampaignCalendarValidationError,
) -> NoReturn:
    if isinstance(exc, CampaignCalendarAuthorizationError):
        _raise_capability_denial(exc.reason)
    raise _bad_request(str(exc)) from exc


def _event_types(values: list[str] | None) -> tuple[str, ...] | None:
    if values is None:
        return None
    event_types = tuple(
        event_type.strip()
        for value in values
        for event_type in value.split(",")
        if event_type.strip()
    )
    return event_types or None


def _event_response(
    event: NormalizedCampaignCalendarEvent,
) -> CampaignCalendarEventResponse:
    data = asdict(event)
    if data["approval"] is not None:
        data["approval"]["available_actions"] = list(
            data["approval"]["available_actions"]
        )
    return CampaignCalendarEventResponse.model_validate(data)


@router.get(
    "/{workspace_id}/campaign-calendar",
    response_model=CampaignCalendarResponse,
)
async def list_campaign_calendar(
    workspace_id: UUID,
    session: SessionDep,
    context: Annotated[CurrentUserContext, Depends(get_current_user_context)],
    start: datetime,
    end: datetime,
    timezone: str = "UTC",
    campaign_id: UUID | None = None,
    artist_id: UUID | None = None,
    release_id: UUID | None = None,
    status_filter: Annotated[str | None, Query(alias="status")] = None,
    event_types: Annotated[list[str] | None, Query()] = None,
    include_archived: bool = False,
    include_published: bool = False,
    limit: Annotated[int, Query(ge=1, le=MAX_CAMPAIGN_CALENDAR_LIMIT)] = 1000,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> CampaignCalendarResponse:
    try:
        page = await campaign_calendar_service.list_campaign_calendar_events(
            session,
            workspace_id,
            actor=context,
            query=CampaignCalendarEventQuery(
                start=start,
                end=end,
                timezone=timezone,
                campaign_id=campaign_id,
                artist_id=artist_id,
                release_id=release_id,
                event_types=_event_types(event_types),
                statuses=(status_filter,) if status_filter is not None else None,
                include_archived=include_archived,
                include_published=include_published,
            ),
            limit=limit,
            offset=offset,
        )
    except (CampaignCalendarAuthorizationError, CampaignCalendarValidationError) as exc:
        _service_error(exc)
    return CampaignCalendarResponse(
        workspace_id=workspace_id,
        start=start,
        end=end,
        timezone=timezone,
        events=[_event_response(event) for event in page.events],
        total=page.total,
        limit=page.limit,
        offset=page.offset,
    )
